### PR TITLE
Implemented recycler view to show query results

### PR DESCRIPTION
**Pull request by Faka-GLB**
Implemented a simple recycler view to show characters names and a thumbnail using Glide.
API is now called in the start of the app, the "Get" button was repurposed to "Try again" in case of error 

Nexus 5X:
image

Nexus S:
image

New loading state
image

New error state
image

Demo:
https://user-images.githubusercontent.com/84337198/132017647-5f38f5ce-56db-46df-a53c-573626aad48d.mp4

Error demo:
https://user-images.githubusercontent.com/84337198/132017675-98a68318-ae92-4199-a005-33488d215058.mp4

App module:
image

Data module:
image

Di and Domain modules:
image

Gradle scripts:
image

### CODE REVIEW

**Review by SartiMau — CHANGES_REQUESTED** [review]
Please use Glide to show the character image in each card pls

**Review by BQuinteros — APPROVED** [review]

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt [hunk 1]
```diff
@@ -42,27 +45,28 @@ class MainActivity : AppCompatActivity() {
...
-        binding.textViewMainActivityWelcomeMessage.visibility = View.VISIBLE
+    private fun successState(response: List<MarvelCharacter>) {
         binding.progressBarMainActivity.visibility = View.GONE
-        makeToast(message = getString(R.string.main_activity_toast_get_characters, response), Toast.LENGTH_LONG)
+        binding.recyclerViewMainActivity.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
+        binding.recyclerViewMainActivity.adapter = CharacterAdapter(response)
+        binding.buttonMainActivityGetInfo.visibility = View.GONE
+        binding.textViewMainActivityWelcomeMessage.visibility = View.GONE
+        binding.recyclerViewMainActivity.visibility = View.VISIBLE
+        binding.textViewMainActivityTitle.visibility = View.GONE
```
Comment (mauroLonghi) [inline app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt:61]: I think that its better first hide the component and then set and put visible the recycler, to be consistent with the hive progress bar in the line 55

File: app/src/main/res/layout/activity_main.xml [hunk 2]
```diff
@@ -44,4 +44,14 @@
...
         android:visibility="gone" />
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/recycler_view_main_activity"
+        style="@style/RecyclerViewStyleMainActivity"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:visibility="gone"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/text_view_main_activity_title"
```
Comment (mauroLonghi) [inline app/src/main/res/layout/activity_main.xml:55]: pls remove "+" in the contranints

File: app/src/main/res/layout/item_character.xml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_height="match_parent">
+
+            <TextView
+                android:id="@+id/text_view_item_character_name"
+                style="@style/TextStyleItemCharacterText"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                app:layout_constraintStart_toStartOf="parent"
```
Comment (mauroLonghi) [inline app/src/main/res/layout/item_character.xml:29]: if you are using match_parent, its not necesary this contraint

Comment (Faka-GLB) [reply app/src/main/res/layout/item_character.xml]: removed

File: app/src/main/res/layout/activity_main.xml (hunk 2 above)
Comment (Faka-GLB) [reply app/src/main/res/layout/activity_main.xml]: removed

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt (hunk 1 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt]: rearrenged lines

**Review by SartiMau — COMMENTED** [review]
Please add an EMPTY_LIST estate or something to show in case of the service response with an empty list

**Review by SartiMau — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt [hunk 4]
```diff
@@ -25,51 +28,65 @@ class MainActivity : AppCompatActivity() {
 
         viewModel.getLiveDataCharacters().observe(this, updateUIObserver)
         setListener()
+        viewModel.getCharacters()
     }
 
     private val updateUIObserver = Observer<Event<Data<MarvelData>>> { event ->
-        when (event?.getContentIfNotHandled()?.responseType) {
+        when (event?.peekContent()?.responseType) {
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt:35]: please don't force the read of the event

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt [hunk 5]
```diff
@@ -25,51 +28,65 @@ class MainActivity : AppCompatActivity() {
...
-    private fun successState(response: Int) {
-        binding.buttonMainActivityGetInfo.visibility = View.VISIBLE
-        binding.textViewMainActivityWelcomeMessage.visibility = View.VISIBLE
+    private fun successState(response: List<MarvelCharacter>) {
+        binding.recyclerViewMainActivity.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
+        binding.recyclerViewMainActivity.adapter = CharacterAdapter(response)
+        binding.textViewMainActivityTitle.visibility = View.GONE
+        binding.textViewMainActivityWelcomeMessage.visibility = View.GONE
+        binding.buttonMainActivityTryAgain.visibility = View.GONE
         binding.progressBarMainActivity.visibility = View.GONE
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt:64]: to show or hide multiple view at once you can try to use constraint layout groups

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt [hunk 6]
```diff
@@ -25,51 +28,65 @@ class MainActivity : AppCompatActivity() {
...
+        binding.recyclerViewMainActivity.visibility = View.VISIBLE
     }
 
     private fun errorState() {
-        binding.buttonMainActivityGetInfo.visibility = View.VISIBLE
-        binding.textViewMainActivityWelcomeMessage.visibility = View.VISIBLE
         binding.progressBarMainActivity.visibility = View.GONE
-        makeToast(message = getString(R.string.main_activity_toast_get_characters_error), Toast.LENGTH_SHORT)
+        binding.textViewMainActivityWelcomeMessage.visibility = View.GONE
+        binding.recyclerViewMainActivity.visibility = View.INVISIBLE
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt:71]: why do you make this invisible and not gone?

File: app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import androidx.recyclerview.widget.RecyclerView
+import com.bumptech.glide.Glide
+import com.onboarding.domain.entity.MarvelCharacter
+import com.onboarding.marvel_mvvm.R
+import com.onboarding.marvel_mvvm.databinding.ItemCharacterBinding
+
+class CharacterAdapter(private val characterItems: List<MarvelCharacter>):  RecyclerView.Adapter<CharacterAdapter.ViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_character, parent, false))
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt:15]: replace return for =

File: app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(characterItems[position])
+    }
+
+    override fun getItemCount(): Int {
+        return characterItems.size
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt:23]: replace return for =

File: app/src/main/res/layout/activity_main.xml [hunk 9]
```diff
@@ -23,25 +23,38 @@
...
-        app:layout_constraintTop_toTopOf="parent" />
+        app:layout_constraintTop_toTopOf="parent"
+        android:visibility="gone" />
 
     <ProgressBar
         android:id="@+id/progress_bar_main_activity"
         android:layout_width="match_parent"
-        android:layout_height="match_parent"
-        android:visibility="gone" />
+        android:layout_height="400dp"
```
Comment (nicolbidone) [inline app/src/main/res/layout/activity_main.xml:46]: please avoid hardcodes

**Review by nicolbidone — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/activity_main.xml (hunk 9 above)
Comment (Faka-GLB) [reply app/src/main/res/layout/activity_main.xml]: avoided

File: app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt (hunk 8 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt]: done

File: app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt (hunk 7 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/adapter/CharacterAdapter.kt]: done

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt (hunk 6 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt]: replaced

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt (hunk 5 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt]: done

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt (hunk 4 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt]: done

File: app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt (hunk 5 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/marvel_mvvm/activity/MainActivity.kt]: redone

**Review by nicolbidone — APPROVED** [review]

**Pull request merged**